### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by Eshy10**
##Game Logic for Tic-Tac-Toe Game
- Added Validation for winning the game
- Added a method to check if the cell is full
- Added a method to check for a valid input
- Added a method to check for user input

### CODE REVIEW

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hi @Eshy10, @vzdrizhni. You have done a good job:+1:. Let's make the following changes:fast_forward:**
1. main.rb should handle only the flow of the game. Let us move interface into a separate class.

2. Also, in the spirit of OOP, let us make player as a separate class. Instances of the player class can be initiated at the start of the game.

3. All classes should be in the /lib directory. You should have a separate file for each class.

4. There should be no **puts, print or get.chomps** outside main.rb. You should refactor your class methods in a way that all puts, print and get.chomps are requested from main.rb.

5. It is better to use **_nil_** to represent initial state of board positions instead of **''_"**. Your display_board method can now render what you want displayed.

6. **update_input** method in main.rb is not called anywhere. If it is not in use, then it should be removed.

_**Kindly make these changes and submit another code review request.**_

**Any question? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Progress :+1: 
You have made some changes as suggested by the previous reviewer but we still have a long way to go before we can merge this PR.
- [ ] `puts` and `p` should only be in the `main.rb` file and nowhere else. This was already pointed out but you never made the change
- [ ] check more comments in for individual sections of your code

### Martin Afani

File: lib/game.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    if (arr[0] == arr[1] && arr[1] == arr[2]) && arr[0..2].none?('_')
+      puts "#{player_name} won on first line"
+      true
+    elsif (arr[3] == arr[4] && arr[4] == arr[5]) && arr[3..5].none?('_')
+      puts "#{player_name} won on the second line"
+      true
+    elsif (arr[6] == arr[7] && arr[7] == arr[8]) && arr[6..8].none?('_')
+      puts "#{player_name} won on the third line"
+      true
+    elsif (arr[0] == arr[3] && arr[3] == arr[6]) && !arr.values_at(0, 3, 6).include?('_')
```
Comment (whiz25) [inline lib/game.rb]: - [ ]  this not an acceptable way for checking for the winners
You can define methods to handle that:
```ruby
def check_win(player_symbol)
    # for rows
    all_equal(board[0..2], player_symbol) ||
    all_equal(board[3..5], player_symbol) ||
    all_equal(board[6..8], player_symbol)
  end

  def all_equal(subarray, player_symbol)
    return true if subarray.uniq.size == 1 && subarray.uniq == player_symbol
  end
```

Comment (Eshy10) [reply lib/game.rb]: I really don't understand you tho, is there a fixed method to check for winning condition?.. please kindly explain. I already remove puts from the game class and it is working perfectly.

Comment (whiz25) [reply lib/game.rb]: @Eshy10 I am not trying to imply that my suggestion is the only way that works. Surely your way also does the job but it isn't the better way to handle that logic. You have a lot of if/else blocks in your code. I have only advised an alternative that can make your code cleaner.

Comment (Eshy10) [reply lib/game.rb]: okay now I understand, I will study your code and look for a way to implement it. thanks

**Review by kobiyoyo — APPROVED** [review]
# Project Approved :thumbsup:
## Hello Team ,  You have done a great job with this project. :clap:

**Pull request merged**